src/referenced.cc

```diff
@@ -0,0 +1,166 @@
+                          << " >= " << global_words.size();
+        return false;
+      }
+      name = global_words[idx];
+    } else {
+      // per-message index is negative, its format is:
+      //    per_message_idx = -(array_idx + 1)
+      idx = -idx - 1;
+      if (idx >= reference.words_size()) {
+        GOOGLE_LOG(ERROR) << "Per message word index is too big: " << idx
+                          << " >= " << reference.words_size();
+        return false;
+      }
+      name = reference.words(idx);
+    }
+
+    if (match.condition() == ReferencedAttributes::ABSENCE) {
+      absence_keys_.push_back(name);
+    } else if (match.condition() == ReferencedAttributes::EXACT) {
+      exact_keys_.push_back(name);
+    }
+  }
+
+  return true;
+}
+
+bool Referenced::Signature(const Attributes& attributes,
+                           std::string* signature) const {
+  for (const std::string& key : absence_keys_) {
+    // if an "absence" key exists, return false for mis-match.
+    if (attributes.attributes.find(key) != attributes.attributes.end()) {
+      return false;
+    }
+  }
+
+  MD5 hasher;
+  for (const std::string& key : exact_keys_) {
+    const auto it = attributes.attributes.find(key);
+    // if an "exact" attribute not present, return false for mismatch.
+    if (it == attributes.attributes.end()) {
+      return false;
+    }
+
+    hasher.Update(it->first);
+    hasher.Update(kDelimiter, kDelimiterLength);
+    switch (it->second.type) {
+      case Attributes::Value::ValueType::STRING:
+        hasher.Update(it->second.str_v);
+        break;
+      case Attributes::Value::ValueType::BYTES:
+        hasher.Update(it->second.str_v);
+        break;
+      case Attributes::Value::ValueType::INT64:
+        hasher.Update(&it->second.value.int64_v,
+                      sizeof(it->second.value.int64_v));
+        break;
+      case Attributes::Value::ValueType::DOUBLE:
+        hasher.Update(&it->second.value.double_v,
+                      sizeof(it->second.value.double_v));
+        break;
+      case Attributes::Value::ValueType::BOOL:
+        hasher.Update(&it->second.value.bool_v,
+                      sizeof(it->second.value.bool_v));
+        break;
+      case Attributes::Value::ValueType::TIME:
+        hasher.Update(&it->second.time_v, sizeof(it->second.time_v));
+        break;
+      case Attributes::Value::ValueType::DURATION:
+        hasher.Update(&it->second.duration_nanos_v,
+                      sizeof(it->second.duration_nanos_v));
+        break;
+      case Attributes::Value::ValueType::STRING_MAP:
+        for (const auto& sub_it : it->second.string_map_v) {
+          hasher.Update(sub_it.first);
+          hasher.Update(kDelimiter, kDelimiterLength);
+          hasher.Update(sub_it.second);
+          hasher.Update(kDelimiter, kDelimiterLength);
+        }
+        break;
+    }
+    hasher.Update(kDelimiter, kDelimiterLength);
+  }
+
+  *signature = hasher.Digest();
```

Comment:
I definitely worry about the amount of work we're doing here in order to produce this digest. That's a lot of cycles for each request. I'll be very interested to see performance profiles for this stuff in the coming months as we start to focus on perf.

In the case of a cache miss, do we end up calculating this digest twice? Once before the cache lookup and once at insertion into the cache? Can we at reuse the digest in that case.

Reply:
Yes, for cache miss, we have calculate N digests if there are N sets of ReferencedAttributes. 
If we want to re-use them in SetResponse,  we may have to save all N digests since we don't know the response ReferencedAttributes.  It is not easy since there are con-current Check requests.

We can leave this as further optimization in the future.